gnsi/acltrie/trie.go

```diff
@@ -0,0 +1,200 @@
+type policies map[pathzpb.Mode]map[string]pathzpb.Action
+
+// insert adds a principal for a given mode and action to policies, returning an error on duplicate entry.
+func (p policies) insert(principal string, mode pathzpb.Mode, action pathzpb.Action) error {
+	m, ok := p[mode]
+	if !ok {
+		p[mode] = make(map[string]pathzpb.Action)
+		m = p[mode]
+	}
+	if _, ok := m[principal]; ok {
+		return fmt.Errorf("policy already contains action for principal")
+	}
+	m[principal] = action
+
+	return nil
+}
+
+// Insert inserts a new policy into the trie.
+func (t *Trie) Insert(r *pathzpb.AuthorizationRule) error {
+	if t.root == nil {
+		t.root = &trieNode{}
+	}
+	if r.Action == pathzpb.Action_ACTION_UNSPECIFIED {
+		return fmt.Errorf("action unspecified")
+	}
+	if r.Mode == pathzpb.Mode_MODE_UNSPECIFIED {
+		return fmt.Errorf("mode unspecified")
+	}
+	if r.GetGroup() == "" && r.GetUser() == "" {
+		return fmt.Errorf("principal unset")
+	}
+
+	path := r.GetPath()
+	node := t.root
+
+	for i, elem := range path.Elem {
+		if elem.Name == "*" {
+			return fmt.Errorf("wildcard path names are not permitted")
+		}
+
+		// Normalize path string by pruning wildcard keys.
+		pathStr, err := elemToString(elem.Name, elem.Key)
+		if err != nil {
+			return fmt.Errorf("invalid path element: %v", err)
+		}
+
+		// The rank of the child is the number of non-wildcard keys (if any).
+		rank := len(setKeys(elem.Key))
+		if rank >= len(node.childrenByRank) { // Resize the slice to [0, rank].
+			node.childrenByRank = append(node.childrenByRank, make([]map[string]*trieNode, rank-len(node.childrenByRank)+1)...)
+		}
+		if node.childrenByRank[rank] == nil {
+			node.childrenByRank[rank] = make(map[string]*trieNode)
+		}
+		children := node.childrenByRank[rank]
+
+		// If the node already exists, keep going.
+		if _, ok := children[pathStr]; ok {
+			node = children[pathStr]
+			continue
+		}
+
+		// Before adding a new node, check if the path conflicts with another policy.
+		for _, child := range children {
+			if checkPathConflict(child.elem, elem) {
+				return fmt.Errorf("policy path conflict with %v/%v", path.Elem[0:i], child.elem)
+			}
+		}
+
+		children[pathStr] = &trieNode{
+			elem:   elem,
+			groups: policies{},
+			users:  policies{},
+		}
+
+		node = children[pathStr]
+	}
+	principal := r.GetUser()
+	policy := node.users
+	if _, isUser := r.GetPrincipal().(*pathzpb.AuthorizationRule_User); !isUser {
+		principal = r.GetGroup()
+		policy = node.groups
+	}
+
+	if err := policy.insert(principal, r.GetMode(), r.GetAction()); err != nil {
+		return fmt.Errorf("error inserting policy at %v: %v", r.Path, err)
+	}
+	node.hasPolicy = true
+
+	return nil
+}
+
+// checkPathConflict returns false if the candidate path overlaps with the accepted path.
+// The candidate path conflicts with the accepted path if its name matches and
+// all keys match (where * matches with anything).
+func checkPathConflict(accepted, candidate *gpb.PathElem) bool {
```

Comment:
A nit is `accepted, candidate` sounds as if there is a difference between them. Would `a, b` just as accurate? Does conflict detection depend on the order of the paths?

Reply:
i found it is easier to reason about the func this way. if you have a strong opinion i can change it